File: 0008. String to Integer (atoi)/Solution.py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = 0
        res = 0
        sign = 1
        n = len(s)

        INT_MAX = 2**31 - 1
        INT_MIN = -2**31

        while i<n and s[i]==' ':
            i+=1
        
        if i<n and (s[i]=='+' or s[i]=='-'):
            if s[i]=='-':
                sign = -1
            i+=1

        while i<n and s[i].isdigit():
            digit = int(s[i])
            
            if res > (INT_MAX - digit) // 10:
                return INT_MAX if sign == 1 else INT_MIN

            res = res*10 + digit
            i+=1

        return sign*res

#using recursion 
# class Solution(object):
#     def myAtoi(self, s):
#         """
#         :type s: str
#         :rtype: int
#         """
#         s = s.strip()
#         if not s:
#             return 0

#         sign = 1
#         start_idx = 0

#         if s[0] == '-':
#             sign = -1
#             start_idx = 1
#         elif s[0] == '+':
#             start_idx = 1

#         return int(self.rec_helper(s, start_idx, 0, sign))

#     def rec_helper(self, s, index, current_res, sign):
#         if index >= len(s) or not s[index].isdigit():
#             return current_res * sign

#         digit = ord(s[index]) - ord('0')
#         current_res = current_res * 10 + digit

#         if sign == 1 and current_res > 2**31 - 1:
#             return 2**31 - 1
#         if sign == -1 and -current_res < -2**31:
#             return -2**31

#         return self.rec_helper(s, index + 1, current_res, sign)

File: 0008. String to Integer (atoi)/Solution.java
class Solution {
    public int myAtoi(String s) {
        int res = 0;
        int n = s.length();
        int i = 0;
        int sign = 1;

        while(s.charAt(i)==' '){
            i++;
        }

        if(i<n &&(s.charAt(i)=='+' || s.charAt(i)=='-')){
            if(s.charAt(i)=='-'){
                sign = -1;
            }
            i++;
        }

        while(i<n && Character.isDigit(s.charAt(i))){
            int digit = s.charAt(i)-'0';
            if(res > (Integer.MAX_VALUE - digit) / 10){
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            res = res*10 + digit;
            i++;
        }

        return sign*res;
    }
}

//using recursion

// class Solution {
//     public int myAtoi(String s) {
//         s = s.trim();
//         if (s.isEmpty()) return 0;

//         int sign = 1;
//         int startIdx = 0;

//         if (s.charAt(0) == '-') {
//             sign = -1;
//             startIdx = 1;
//         } else if (s.charAt(0) == '+') {
//             startIdx = 1;
//         }

//         return (int) recHelper(s, startIdx, 0L, sign);
//     }

//     private long recHelper(String s, int index, long currentRes, int sign) {
//         if (index >= s.length() || !Character.isDigit(s.charAt(index))) {
//             return currentRes * sign;
//         }

//         int digit = s.charAt(index) - '0';
//         currentRes = currentRes * 10 + digit;

//         if (sign == 1 && currentRes > Integer.MAX_VALUE) return Integer.MAX_VALUE;
//         if (sign == -1 && (-currentRes) < Integer.MIN_VALUE) return Integer.MIN_VALUE;

//         return recHelper(s, index + 1, currentRes, sign);
//     }
// }